src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SQLDNSCache.cs

```diff
@@ -0,0 +1,87 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.Data.SqlClient
+{
+    internal class SQLDNSCache
+    {
+        private static readonly SQLDNSCache _SQLDNSCache = new SQLDNSCache();
+        private static readonly int initialCapacity = 100;
+        private ConcurrentDictionary<string, SQLDNSInfo> DNSInfoCache;
+
+        // singleton instance
+        public static SQLDNSCache Instance { get { return _SQLDNSCache; } }
+
+        private SQLDNSCache()
+        {
+            int level = 4 * Environment.ProcessorCount;
+            DNSInfoCache = new ConcurrentDictionary<string, SQLDNSInfo>(concurrencyLevel: level,
+                                                                            capacity: initialCapacity,
+                                                                            comparer: StringComparer.OrdinalIgnoreCase);
+        }
+
+        internal bool AddDNSInfo(SQLDNSInfo item)
+        {
+            if (null != item)
+            {
+                if (DNSInfoCache.ContainsKey(item.FQDN))
+                {
+
+                    DeleteDNSInfo(item.FQDN);
+                }
+
+                return DNSInfoCache.TryAdd(item.FQDN, item);
```

Comment:
Could you use `AddorUpdate` instead of deleting first?

Reply:
I was trying to use the AddorUpate method. But from the [documents](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=netcore-3.1), it mentions 
>If you call AddOrUpdate simultaneously on different threads, addValueFactory may be called multiple times, but its key/value pair might not be added to the dictionary for every call.
>
>For modifications and write operations to the dictionary, ConcurrentDictionary<TKey,TValue> uses fine-grained locking to ensure thread safety. (Read operations on the dictionary are performed in a lock-free manner.) However, the addValueFactory and updateValueFactory delegates are called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, AddOrUpdate is not atomic with regards to all other operations on the ConcurrentDictionary<TKey,TValue> class.

And it requires addition function for updating the records so I think remove and add the value again will be safer and straight-forward.